### ISSUE TITLE
PKI: new certificates valid ahead of time

### ISSUE THREAD

**Issue by gautaz** [hashicorp/vault#4639]
**Is your feature request related to a problem? Please describe.**
Sometimes the peer that your trying to connect to is lagging. When using certificates just created for the purpose of the connection, the peer sees the new certificate as valid "in the future" and thus rejects it for now.
A way out is to ensure that the peer is synchronized (NTP) but this can be hard to achieve when the peer is newly installed (think embedded) and only accepts new settings (like the NTP server) once a valid certificate is provided.
In this case an operator is asked to manually set the device's clock. As this operator cannot be expected to always achieve perfect timing one way or another, having a way to produce certificates that are already valid since a few minutes might mitigate this type of issue.

**Describe the solution you'd like**
An option when creating/updating a PKI role to extend the validity period of each new certificate for a few minutes back in the past might do the job.

**Describe alternatives you've considered**
For now the only way out is to ask the operator to care about setting the device ahead of time.

**Explain any additional use-cases**
No other use case I'm thinking of right now...

**Comment by robison** [comment]
Per [this](https://github.com/hashicorp/vault/blob/187c051ef34872beed574c8c30580c8f28c538c4/builtin/logical/pki/cert_util.go#L1216) bit of code, NotBefore is already set to curTime - 30s.

**Comment by gautaz** [comment]
The error margin for an operator entering manually the time can easily exceed 30s.

On large deployments this leads to many round-trips between the support an the client sites dealing directly with the devices. Being able to extend this delay by making it configurable (at least for a few minutes) would greatly mitigate this type of issue when necessary.

**Comment by robison** [comment]
Not a committer, but I'd say that's not unreasonable. Let's Encrypt backdates certificates by 1 hour to account for this, and that's for 90 day (2160 hour) certs. I think that the intent with vault is to limit cert lifetimes as much as possible; would it make sense to do something like:

```
	preIssueNotBefore := data.params.NotAfter.Sub(time.Now()) / (2160 * time.Second)
	if preIssueNotBefore < (30 * time.Second) {
		preIssueNotBefore = 30 * time.Second
	}
	...
	...
	NotBefore: time.Now().Sub(preIssueNotBefore),
```

Alternatively, setting this as configurable on the role might also be an option.

**Comment by gautaz** [comment]
@robison I like your idea of a computed pre-issue time but the current proposal means that the pre-issue time would start to exceed 30s once the lifetime of the certificate exceeds 18h if my math is correct.

As per example, this won't do with my current use case which needs a temporary certificate (lets say 15 minutes) for a temporary HTTPS server that the device will connect on to retrieve some configuration data. 

I guess it will be difficult to find an algorithm to compute the pre-issue time that will match all needs, perhaps some sort of logarithmic scaling starting at 30s for values that exceeds 5 minutes but that seems so awkward that I would rather let the user decide.

I do not even know if some safeguards must exist to avoid having a user put values like 1 year for the pre-issue time.

**Comment by robison** [comment]
I daresay that if you need a certificate with such a short lifetime, and that the network time variance in your infrastructure is high enough to require more than a 30-second pre-issuance `NotBefore`,  that effort may be better spent lowering that time variance.

Presumably that's something you've already thought about, and may not be able to accomplish - I'm simply stipulating it for purposes of the conversation here. In this particular case, I think the only way to solve it would be either extending the code and using a role setting to accomplish this, or perhaps modifying your internal build to increase that 30 second timer.

**Comment by gautaz** [comment]
Fair enough :-).

Would a PR proposal for a role setting do?
If so I will ask for resources internally to make this happen.

**Comment by jefferai** [comment]
Sure.

**Comment by gautaz** [comment]
Just some follow-up.
I have asked for resources internally. For now it is waiting to be scheduled in the project's timeframe.

**Comment by catsby** [comment]
Hello - after a long delay, I'm closing this issue as it was implemented in #5325. Thanks!
